### ISSUE TITLE
Is it possible to use ImageDataGenerator within Keras Tuner?

### ISSUE THREAD

**Issue by EgorFollia** [keras-team/keras-tuner#223]
I am using `from tensorflow.keras.preprocessing.image import ImageDataGenerator`
Here is the code :
```
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
x_train, x_test = x_train.astype('float32') / 255, x_test.astype('float32') / 255
y_train, y_test = to_categorical(y_train), to_categorical(y_test)

img_gen = ImageDataGenerator(
    rotation_range=10,
    width_shift_range=.1,
    height_shift_range=.1,
    horizontal_flip=True,
    vertical_flip=True,
    fill_mode="nearest"
)

img_gen_train = img_gen.flow(x_train, y_train, batch_size=128, shuffle=True)

rand_tuner = RandomSearch(hypermodel=hypermodel,
                          objective='val_acc',
                          max_trials=max_trials,
                          project_name='cifar10')
```

then I am building a model using keras tuner HyperModel:
```
def hypermodel(hp):
    units_choice = hp.Int('units', min_value=32, max_value=512, step=32, default=128)
    lr_choice = hp.Float('learning_rate', 1e-5, 1e-2, sampling='LOG', default=1e-3)
    dropout_rate_choice = hp.Float('rate', 0, .5, step=.1, default=.2)
    filters_choice = hp.Choice('num_filters', values=[32, 64], default=64)

    model = Sequential()
    model.add(Conv2D(filters=16, kernel_size=3, 
                            activation='relu', input_shape=input_shape))
    model.add(Dropout(rate=dropout_rate_choice))
    model.add(Flatten())
    model.add(Dense(units=units_choice, activation='relu'))
    model.add(Dense(num_labels, activation='softmax'))
    model.compile(loss='sparse_categorical_crossentropy',
                optimizer=Adam(lr_choice),
                metrics=['accuracy'])

    return model
```

After that, in an attempt to search best got an error : 
[![enter image description here][1]][1]

  [1]: https://i.stack.imgur.com/YlVDr.png

**How to deal with it?**

**Comment by omalleyt12** [comment]
@EgorFollia Thanks for the issue!

Yes, it's possible to use `ImageDataGenerator` with Keras Tuner

You should either remove `to_categorical` or use `categorical_crossentropy` instead of `spase_categorical_crossentropy`

`to_categorical` converts your labels to one-hot encoding. "sparse_categorical_crossentropy" expects the labels to be scalar ints

**Comment by rodrigocardoso92** [comment]
Hello @EgorFollia,

You need to declare the generator in your hypermodel function.

**Comment by 21kc-caracol** [comment]
is it possible to combine `batch_size` tuning when I use an `ImageDataGenerator`?
@rodrigocardoso92 | @omalleyt12

Could you assist with a code snippet? The example in this [similar issue](https://github.com/keras-team/keras-tuner/issues/122)
didn't help.

**Comment by Ashgreene** [comment]
@21kc-caracol 

This is late, but I was able to combine `batch_size` tuning with the `ImageDataGenerator`. Maybe it'll save someone else some time. Here's an example how I did it:

```
class MyHyperModel(kt.HyperModel):

    def build(self, hp): # Use hp for hyperparameter definitions
        
        # Define parameters to tune
        optimizer_ops = ['adam', 'SGD', 'AdamW', 'RMSProp']
        optimizer = hp.Choice('optimizer', optimizer_ops)
        l2_reg = hp.Float('l2_reg', min_value=1e-5, max_value=1e-2, sampling='log')
        dropout = hp.Float('dropout', min_value=0.0, max_value=0.5, step=0.05)

        # Building a CNN model
        model = Sequential()
        model.add(Conv2D(hp.Int("conv_1", min_value=16, max_value=128, step=16), (3, 3), padding='same',
                     input_shape=x_train.shape[1:], 
                     kernel_regularizer=regularizers.l2(l2_reg)))
        model.add(Activation('relu'))
        model.add(MaxPooling2D(pool_size=(2, 2)))
        model.add(Dropout(0.25))

        model.add(Flatten())
        model.add(Dense(hp.Int("dense_1", min_value=64, max_value=512, step=8), 
                        kernel_regularizer=regularizers.l2(l2_reg)))
        model.add(Dense(hp.Int("dense_2", min_value=64, max_value=512, step=8), 
                        kernel_regularizer=regularizers.l2(l2_reg)))
        model.add(Activation('relu'))
        model.add(Dropout(dropout))
        model.add(Dense(2))
        model.add(Activation('sigmoid'))

    # Compile model
        model.compile(loss='categorical_crossentropy',
                    optimizer=optimizer,
                    metrics=['accuracy'])
        return model

# Incorporate ImageDataGenerator code into your fit function
def fit(self, hp, model, *args, **kwargs):
        batch_size=hp.Choice("batch_size", [5, 8, 10, 16, 20])

        # creating image data generator to center and augment images
        datagen = ImageDataGenerator(featurewise_center=True, shear_range=0.2,
                             zoom_range=0.2, horizontal_flip=True)

        # calculating the mean of the training dataset
        datagen.fit(x_train)

        # preparing iterators to scale images
        train_iterator = datagen.flow(x_train, y_train, batch_size=batch_size)
        test_iterator = datagen.flow(x_test, y_test, batch_size=batch_size)
      
        return model.fit(
            *args,
            batch_size=batch_size,
            **kwargs,
        )


# Create a Keras Tuner tuner object (e.g., Hyperband)
tuner = kt.Hyperband(
    MyHyperModel(),
    objective='val_accuracy', # Define the objective to optimize
    max_epochs=3, # Number of hyperparameter combinations to try
    seed=42, # Random seed for reproducibility
    factor=3, # Reduction factor for the number of epochs
    overwrite=True, # Overwrite previous results if they exist
    directory='my_dir', # Directory to save results
    project_name='my_project' # Name of the project
)

# Perform hyperparameter search using datasets originally input to ImageDataGenerator. 
#Note that I kept in an EarlyStopping callback not shown here just for example purposes
tuner.search(x_train, y_train,        
             epochs=50,  # Set a fixed number of epochs
             validation_data=(x_test, y_test),
             callbacks=[es])
```

**Comment by Ashgreene** [comment]
EDIT: 

To make sure you can _retrieve_ the batch size, define batch size in the `build` function, e.g.,         `batch_size=hp.Choice("batch_size", [10, 20, 30])
`
Then, in the `fit` function, use `batch_size = hp.get("batch_size")` instead of defining the batch size options.
